kindelia_core/src/net.rs

```diff
@@ -72,23 +87,46 @@ impl std::fmt::Display for Address {
   }
 }
 
+#[derive(Error, Debug)]
+pub enum ParseAddressError {
+  #[error("Invalid IP Address '{value:?}'")]
+  InvalidAddress {
+    value: String,
+    source: ParseIntError,
+  },
+  #[error("Invalid Port '{value:?}'")]
+  InvalidPort {
+    value: String,
+    source: ParseIntError,
+  },
+}
+
 /// Converts a string to an UDP Address.
 /// TODO: UNSAFE.
-pub fn parse_address(code: &str) -> Address {
+pub fn parse_address(code: &str) -> Result<Address, ParseAddressError> {
   let strs = code.split(':').collect::<Vec<&str>>();
-  let vals =
-    strs[0].split('.').map(|o| o.parse::<u8>().unwrap()).collect::<Vec<u8>>();
-  let port = strs.get(1).map(|s| s.parse::<u16>().unwrap()).unwrap_or(UDP_PORT);
-  Address::IPv4 {
+  let vals = strs[0]
+    .split('.')
+    .map(|o| {
+      o.parse::<u8>()
+        .map_err(|e| ParseAddressError::InvalidAddress{value: strs[0].to_string(), source: e}) })
+    .collect::<Result<Vec<u8>, ParseAddressError>>()?;
+  let port = match strs.get(1) {
+    Some(s) => s
+      .parse::<u16>()
+      .map_err(|e| ParseAddressError::InvalidAddress{value: strs[0].to_string(), source: e})?,
+    None => UDP_PORT,
+  };
+  Ok(Address::IPv4 {
     val0: vals[0],
     val1: vals[1],
     val2: vals[2],
     val3: vals[3],
     port,
-  }
+  })
 }
```

Comment:
These `array[n]` type accesses can also cause unwanted `panics`. Doing this code without them however would be pretty boring.

I don't know if there was a reason why this address wasn't created as a `std::net::SocketAddrV4` or a `std::net::SocketAddr` (the latter is an enum that already contains IPv6 socket addresses, which would resolve existing IPv6 TODOs in the code); both contain a `.parse()` method which already does this parsing work for us. I'll talk to @steinerkelvin about it and I think I can change it, but it will greatly increase the scope of this PR if done here (many changes to `node.rs` will be needed).

Maybe we can leave it as it is in this PR, and open a new issue for this to be done.

Reply:
> These array[n] type accesses can also cause unwanted panics. Doing this code without them however would be pretty boring.

great point.  yeah, I think we could deal with these in another pass.  I will add it to my list.   Shouldn't be too bad, something like: 

```
vals.get(0).ok_or(ArrayOutofBounds(0))?
```

which could even become a macro:  `aget!(vals,0)?` or maybe use a trait `vals.eget(0)?`.  Maybe add some context.   Likely there is already a crate for this if we look.

btw, I found there is a crate [no_panic](https://docs.rs/no-panic/latest/no_panic/) that provides a macro which forces compiler to verify that a function *cannot* panic and provides that guarantee to callers.  Unfortunately many std lib calls panic, so it can't be used everywhere.  I'll play around with it also.  I'm not certain yet how much value it can provide us.

> std::net::SocketAddr

yeah, I've been thinking it would be best to use that, or possibly find a crate with a higher level type that accepts either domain:port or ip:port (v4 or v6) and resolves domains securely, at least for accepting cli/config args.   

> Maybe we can leave it as it is in this PR, and open a new issue for this to be done

yeah, definitely something for a separate PR.  outside the scope here.